File: lib/codeintel/autoindex/inference/typescript.go
package inference

import (
	"context"
	"encoding/json"
	"path/filepath"

	"github.com/grafana/regexp"

	"github.com/sourcegraph/sourcegraph/lib/codeintel/autoindex/config"
)

func TypeScriptPatterns() []*regexp.Regexp {
	return []*regexp.Regexp{
		pathPattern(rawPattern("tsconfig.json")),
		pathPattern(rawPattern("package.json")),
		pathPattern(rawPattern("lerna.json")),
		pathPattern(rawPattern("yarn.lock")),
		pathPattern(rawPattern(".nvmrc")),
		pathPattern(rawPattern(".node-version")),
		pathPattern(rawPattern(".n-node-version")),
	}
}

const lsifTypescriptImage = "sourcegraph/lsif-typescript:autoindex"
const nMuslCommand = "N_NODE_MIRROR=https://unofficial-builds.nodejs.org/download/release n --arch x64-musl auto"

var tscSegmentBlockList = append([]string{"node_modules"}, segmentBlockList...)

func inferSingleTypeScriptIndexJob(
	gitclient GitClient, pathMap pathMap, tsConfigPath string, shouldInferConfig bool,
) *config.IndexJob {
	if !containsNoSegments(tsConfigPath, tscSegmentBlockList...) {
		return nil
	}
	isYarn := checkLernaFile(gitclient, tsConfigPath, pathMap)
	var dockerSteps []config.DockerStep
	for _, dir := range ancestorDirs(tsConfigPath) {
		if !pathMap.contains(dir, "package.json") {
			continue
		}

		ignoreScripts := ""
		if shouldInferConfig {
			ignoreScripts = " --ignore-scripts"
		}
		var commands []string
		if isYarn || pathMap.contains(dir, "yarn.lock") {
			commands = append(commands, "yarn --ignore-engines"+ignoreScripts)
		} else {
			commands = append(commands, "npm install"+ignoreScripts)
		}

		dockerSteps = append(dockerSteps, config.DockerStep{
			Root:     dir,
			Image:    lsifTypescriptImage,
			Commands: commands,
		})
	}

	var localSteps []string
	if checkCanDeriveNodeVersion(gitclient, tsConfigPath, pathMap) {
		for i, step := range dockerSteps {
			step.Commands = append([]string{nMuslCommand}, step.Commands...)
			dockerSteps[i] = step
		}

		localSteps = append(localSteps, nMuslCommand)
	}

	n := len(dockerSteps)
	for i := 0; i < n/2; i++ {
		dockerSteps[i], dockerSteps[n-i-1] = dockerSteps[n-i-1], dockerSteps[i]
	}

	indexerArgs := []string{"lsif-typescript-autoindex", "index"}
	if shouldInferConfig {
		indexerArgs = append(indexerArgs, "--infer-tsconfig")
	}

	return &config.IndexJob{
		Steps:       dockerSteps,
		LocalSteps:  localSteps,
		Root:        dirWithoutDot(tsConfigPath),
		Indexer:     lsifTypescriptImage,
		IndexerArgs: indexerArgs,
		Outfile:     "",
	}
}

func InferTypeScriptIndexJobs(gitclient GitClient, paths []string) (indexes []config.IndexJob) {
	pathMap := newPathMap(paths)

	tsConfigEntry, tsConfigPresent := pathMap["tsconfig.json"]
	if !tsConfigPresent {
		indexJob := inferSingleTypeScriptIndexJob(gitclient, pathMap, "tsconfig.json", true)
		if indexJob != nil {
			return []config.IndexJob{*indexJob}
		}
		return []config.IndexJob{}
	}

	for _, tsConfigIndex := range tsConfigEntry.indexes {
		indexJob := inferSingleTypeScriptIndexJob(gitclient, pathMap, paths[tsConfigIndex], false)
		if indexJob != nil {
			indexes = append(indexes, *indexJob)
		}
	}

	return indexes
}

func checkLernaFile(gitclient GitClient, path string, pathMap pathMap) (isYarn bool) {
	lernaConfig := struct {
		NpmClient string `json:"npmClient"`
	}{}

	for _, dir := range ancestorDirs(path) {
		if pathMap.contains(dir, "lerna.json") {
			lernaPath := filepath.Join(dir, "lerna.json")
			if b, err := gitclient.RawContents(context.TODO(), lernaPath); err == nil {
				if err := json.Unmarshal(b, &lernaConfig); err == nil && lernaConfig.NpmClient == "yarn" {
					return true
				}
			}
		}
	}
	return false
}

func checkCanDeriveNodeVersion(gitclient GitClient, path string, pathMap pathMap) bool {
	for _, dir := range ancestorDirs(path) {
		packageJSONPath := filepath.Join(dir, "package.json")
		if (pathMap.contains(dir, "package.json") && hasEnginesField(gitclient, packageJSONPath)) ||
			pathMap.contains(dir, ".nvmrc") ||
			pathMap.contains(dir, ".node-version") ||
			pathMap.contains(dir, ".n-node-version") {
			return true
		}
	}
	return false
}

func hasEnginesField(gitclient GitClient, packageJSONPath string) (hasField bool) {
	packageJSON := struct {
		Engines *struct {
			Node *string `json:"node"`
		} `json:"engines"`
	}{}

	if b, err := gitclient.RawContents(context.TODO(), packageJSONPath); err == nil {
		if err := json.Unmarshal(b, &packageJSON); err == nil {
			if packageJSON.Engines != nil && packageJSON.Engines.Node != nil {
				return true
			}
		}
	}
	return
}

File: internal/codeintel/autoindexing/internal/inference/lua/typescript.lua
local fun = require "fun"
local json = require "json"
local path = require "path"
local pattern = require "sg.autoindex.patterns"
local recognizer = require "sg.autoindex.recognizer"
local shared = require "sg.autoindex.shared"

local indexer = "sourcegraph/scip-typescript:autoindex"
local n_node_mirror = "https://unofficial-builds.nodejs.org/download/release"
local typescript_nmusl_command = "N_NODE_MIRROR=" .. n_node_mirror .. " n --arch x64-musl auto"
local node_derivable_filenames = {
  ".nvmrc",
  ".node-version",
  ".n-node-version",
}

local exclude_paths = pattern.new_path_combine(shared.exclude_paths, {
  pattern.new_path_segment "node_modules",
})

local contains = function(table, element)
  return fun.any(function(v)
    return v == element
  end, table)
end

local contains_any = function(paths, candidates)
  return fun.any(function(v)
    return contains(paths, v)
  end, candidates)
end

local reverse = function(slice)
  local reversed = {}
  for i = #slice, 1, -1 do
    reversed[#reversed + 1] = slice[i]
  end

  return reversed
end

local with_new_head = function(slice, element)
  local new = { element }
  for _, v in ipairs(slice) do
    table.insert(new, v)
  end

  return new
end

local safe_decode = function(encoded)
  local _, payload = pcall(function()
    return json.decode(encoded)
  end)
  return payload
end

local check_lerna_file_contents = function(contents)
  local payload = safe_decode(contents)
  return payload and payload["npmClient"] == "yarn"
end

local check_package_json_contents = function(contents)
  local payload = safe_decode(contents)
  return payload and payload["engines"] and payload["engines"]["node"]
end

local check_lerna_file = function(root, contents_by_path)
  return fun.any(function(a)
    return check_lerna_file_contents(contents_by_path[path.join(a, "lerna.json")] or "")
  end, path.ancestors(root))
end

local can_derive_node_version = function(root, paths, contents_by_path)
  return fun.any(function(a)
    return check_package_json_contents(contents_by_path[path.join(a, "package.json")] or "")
      or contains_any(
        paths,
        fun.map(function(filename)
          return path.join(a, filename)
        end, node_derivable_filenames)
      )
  end, path.ancestors(root))
end

local infer_typescript_job = function(api, tsconfig_path, should_infer_config)
  local root = path.dirname(tsconfig_path)
  local reverse_ancestors = reverse(path.ancestors(tsconfig_path))

  api:register(recognizer.new_path_recognizer {
    patterns = {
      -- To disambiguate installation steps
      pattern.new_path_basename "yarn.lock",
      -- Try to determine version
      pattern.new_path_basename ".n-node-version",
      pattern.new_path_basename ".node-version",
      pattern.new_path_basename ".nvmrc",
      -- To reinvoke simple cases with no other files
      pattern.new_path_basename "tsconfig.json",
      pattern.new_path_exclude(exclude_paths),
    },

    patterns_for_content = {
      -- To read explicitly configured engines and npm client
      pattern.new_path_basename "package.json",
      pattern.new_path_basename "lerna.json",
      pattern.new_path_exclude(exclude_paths),
    },

    -- Invoked when any of the files listed in the patterns above exist.
    generate = function(api, paths, contents_by_path)
      local is_yarn = check_lerna_file(root, contents_by_path)

      local docker_steps = fun.totable(fun.map(function(ra)
        if contents_by_path[path.join(ra, "package.json")] then
          local install_command = ""
          if is_yarn or contains(paths, path.join(ra, "yarn.lock")) then
            install_command = "yarn --ignore-engines"
          else
            install_command = "npm install"
          end

          local install_command_suffix = ""
          if should_infer_config then
            install_command_suffix = " --ignore-scripts"
          end

          return {
            root = ra,
            image = indexer,
            commands = { install_command .. install_command_suffix },
          }
        end
      end, reverse_ancestors))

      local local_steps = {}
      if can_derive_node_version(root, paths, contents_by_path) then
        docker_steps = fun.totable(fun.map(function(s)
          -- Add `n` invocation command before each docker step
          s.commands = with_new_head(s.commands, typescript_nmusl_command)
          return s
        end, docker_steps))

        -- Add `n` invocation (in indexing container) before indexer runs
        local_steps = { typescript_nmusl_command }
      end

      local args = { "scip-typescript", "index" }
      if should_infer_config then
        table.insert(args, "--infer-tsconfig")
      end

      return {
        steps = docker_steps,
        local_steps = local_steps,
        root = root,
        indexer = indexer,
        indexer_args = args,
        outfile = "index.scip",
      }
    end,
  })

  return {}
end

return recognizer.new_path_recognizer {
  patterns = {
    pattern.new_path_basename "package.json",
    pattern.new_path_basename "tsconfig.json",
    pattern.new_path_exclude(exclude_paths),
  },

  -- Invoked when package.json or tsconfig.json files exist
  generate = function(api, paths)
    local has_tsconfig = false
    fun.each(function(p)
      if path.basename(p) == "tsconfig.json" then
        -- Infer typescript jobs
        infer_typescript_job(api, p, false)
        has_tsconfig = true
      end
    end, paths)

    if not has_tsconfig then
      -- Infer javascript jobs if there's no tsconfig.json found
      infer_typescript_job(api, "tsconfig.json", true)
    end

    return {}
  end,
}
